src/rgw/driver/rados/topic_migration.cc

```diff
@@ -0,0 +1,288 @@
+
+  return 0;
+}
+
+// migrate v1 notification metadata for a single bucket
+int migrate_notification(const DoutPrefixProvider* dpp, optional_yield y,
+                         rgw::sal::RadosStore* driver, const rgw_raw_obj& obj)
+{
+  // parse bucket name and marker of out "pubsub.{tenant}.bucket.{name}/{marker}"
+  auto* rados = driver->getRados()->get_rados_handle();
+  std::string tenant;
+  std::string bucket_name;
+  std::string marker;
+  int r = deconstruct_topics_oid(obj.oid, tenant, bucket_name, marker, dpp);
+  if (r < 0) {
+    std::string s = "failed to read tenant, bucket name and marker from " + obj.to_str()
+                  + " with: " + cpp_strerror(r)
+                  + " expected format pubsub.{tenant}.bucket.{name}/{marker}!";
+    rgw_clog_warn(rados, s);
+    return r;
+  }
+
+  // migrate the notifications
+  rgw_pubsub_bucket_topics v1_bucket_topics;
+  rgw_bucket rgw_bucket_info(tenant, bucket_name);
+  rgw_bucket_info.marker = marker;
+  rgw::sal::RadosBucket rados_bucket(driver, rgw_bucket_info);
+  RGWObjVersionTracker bucket_topics_objv;
+  r = rados_bucket.read_topics(v1_bucket_topics, &bucket_topics_objv, y, dpp);
+  if (r == -ENOENT) {
+    return 0; // ok, someone else already migrated
+  }
+  if (r < 0) {
+    std::string s = "failed to read v1 bucket notifications from " + obj.to_str()
+                  + " with: " + cpp_strerror(r);
+    rgw_clog_warn(rados, s);
+    return r;
+  }
+
+  // in a for-loop that retries ECANCELED errors:
+  // {
+  // load the corresponding bucket by name
+  // break if marker doesn't match loaded bucket's
+  // merge with existing RGW_ATTR_BUCKET_NOTIFICATION topics (don't override existing v2)
+  // write RGW_ATTR_BUCKET_NOTIFICATION xattr
+  // }
+  std::unique_ptr<rgw::sal::Bucket> bucket;
+  r = -ECANCELED;
+  for (auto i = 0u; i < 15u && r == -ECANCELED; ++i) {
+    r = driver->load_bucket(dpp, rgw_bucket_info, &bucket, y);
+    if (r == -ENOENT) {
+      break; // bucket is deleted, we should delete the v1 notification
+    }
+    if (r < 0) {
+      std::string s = "failed to load the bucket from " + obj.to_str()
+                    + " with: " + cpp_strerror(r);
+      rgw_clog_warn(rados, s);
+      return r;
+    }
+
+    if (bucket->get_marker() != marker) {
+      break;
+    }
+
+    rgw::sal::Attrs& attrs = bucket->get_attrs();
+
+    rgw_pubsub_bucket_topics v2_bucket_topics;
+    if (const auto iter = attrs.find(RGW_ATTR_BUCKET_NOTIFICATION); iter != attrs.end()) {
+      // bucket notification v2 already exists
+      try {
+        const auto& bl = iter->second;
+        auto biter = bl.cbegin();
+        v2_bucket_topics.decode(biter);
+      } catch (buffer::error& err) {
+        ldpp_dout(dpp, 1) << "ERROR: failed to decode v2 bucket topics for bucket: "
+                      << bucket->get_name() << dendl;
+        return -EIO;
+      }
+    }
+      
+    v2_bucket_topics.topics.merge(v1_bucket_topics.topics);
```

Comment:
I think it is safer to change the (internal) topic name to the original topic because someone might use it incorrectly
```
    // merge v1 and v2 bucket topics
    for (auto& [topic_name, bucket_topic]: bucket_topics.topics) {
      bucket_topic.topic.name = bucket_topic.topic.dest.arn_topic;
      v2_bucket_topics.topics.insert({topic_name, bucket_topic});
    }
```

Reply:
so, if i use the above instead of "merge(), i won't need the change in:
https://github.com/ceph/ceph/pull/55214/files#diff-e05836df60adf79bd9d629146b894bcd2ac828ea23a63c324e87b20cf7650d72L1089

Reply:
yes